doc/architecture.md

```diff
@@ -0,0 +1,181 @@
+as we might like or imagine it to be. Within Librarian, the term "library" has a very specific
+meaning, namely "an entry within the `pipeline-state.json` file, within a given repository".
+Librarian cares about information relating to a library such as:
+
+- The library ID: this is how Librarian instructs language-specific containers to operate
+  on a particular library
+- The most recently-released version (if any) of the library
+- Which API paths (in the API specification repo) are used to generate code that's contained
+  within the library
+- Which source paths (within the language repo that contains the state file) contribute to
+  the library, so that Librarian can determine that a new version of the library should be released
+
+This is not an exhaustive list; see [pipeline.proto](../proto/pipeline.proto) for more details.
+
+This approach is designed to accommodate the requirements of different languages:
+
+- A library can include code generated from multiple API paths. Different languages package APIs at
+  different granularities. Librarian will 
+- A library doesn't have to contain any generated code at all. A library which only contains handwritten
+  code simply won't have any API paths associated with it.
+- A library doesn't have to correspond to a single package in a package manager. Some languages
+  don't have package managers as such, and some libraries might naturally be comprised of multiple
+  packages.
+- A library isn't absolutely *required* to have any source at all. Currently .NET uses this flexibility
+  to publish API reference documentation for dependencies, along with a product neutral "help" guide.
+  This is an unusual use case though.
+
+Fully-generated libraries are typically configured automatically via the `configure` CLI command
+(which invokes the [`configure` container command](container-contract.md#configure)).
+However, they can also be configured manually - typically for handwritten libraries.
+
+The ID of a library is determined by whatever configures that library initially - either the
+language container or the human performing manual configuration. Librarian treats this as an opaque
+string, without expecting any particular format. These IDs are included in pull request and commit
+messages, so it's useful for them to be human readable, but beyond that different languages have a large
+degree of freedom. If a library is published as a single package (e.g. a single NuGet package or Ruby gem)
+we would recommend using the name of the package as the library ID.
+
+Currently, Librarian does not validate the IDs specified in the state file, but it is likely that
+we will add constraints later. Those constraints are likely to be primarily around characters (e.g.
+only printable characters, potentially only ASCII and prohibiting some characters that could cause
+issues when used as a filename, e.g. `/`) - and an empty library ID is likely to be prohibited.
```

Comment:
Go has pretty good rules for this. See https://pkg.go.dev/golang.org/x/mod@v0.25.0/module#CheckPath.